src/coreclr/vm/amd64/JitHelpers_Fast.asm

```diff
@@ -256,6 +256,243 @@ endif
+        je      Exit
+    SkipCheck:
+
+        cmp     [g_region_use_bitwise_write_barrier], 0
+        je      CheckCardTableByte
+
+        ; compute card table bit
+        mov     rcx, rdi
+        mov     al, 1
+        shr     rcx, 8
+        and     cl, 7
+        shl     al, cl
+
+        ; move current rdi value into rcx and then increment the pointers
+        mov     rcx, rdi
+        add     rsi, 8h
+        add     rdi, 8h
+
+        ; Check if we need to update the card table
+        ; Calc pCardByte
+        shr     rcx, 0Bh
+        add     rcx, [g_card_table]
+
+        ; Check if this card table bit is already set
+        test    byte ptr [rcx], al
+        je      SetCardTableBit
+        dec     r8d
+        jne     NextByref
+        REPRET
+
+    SetCardTableBit:
+        lock or byte ptr [rcx], al
+        jmp     CheckCardBundle
+endif
+CheckCardTableByte:
+
+        ; move current rdi value into rcx and then increment the pointers
+        mov     rcx, rdi
+        add     rsi, 8h
+        add     rdi, 8h
+
+        ; Check if we need to update the card table
+        ; Calc pCardByte
+        shr     rcx, 0Bh
+        add     rcx, [g_card_table]
+
+        ; Check if this card is dirty
+        cmp     byte ptr [rcx], 0FFh
+        jne     UpdateCardTable
+        dec     r8d
+        jne     NextByref
+        REPRET
+
+    UpdateCardTable:
+        mov     byte ptr [rcx], 0FFh
+
+    CheckCardBundle:
+
+ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        ; check if we need to update the card bundle table
+        ; restore destination address from rdi - rdi has been incremented by 8 already
+        lea     rcx, [rdi-8]
+        shr     rcx, 15h
+        add     rcx, [g_card_bundle_table]
+        cmp     byte ptr [rcx], 0FFh
+        jne     UpdateCardBundleTable
+        dec     r8d
+        jne     NextByref
+        REPRET
+
+    UpdateCardBundleTable:
+        mov     byte ptr [rcx], 0FFh
+endif
+        dec     r8d
+        jne     NextByref
+        ret
+
+    align 16
+    NotInHeap:
+; If WRITE_BARRIER_CHECK then we won't have already done the mov and should do it here
+; If !WRITE_BARRIER_CHECK we want _NotInHeap and _Leave to be the same and have both
+; 16 byte aligned.
+ifdef WRITE_BARRIER_CHECK
+        ; rcx is [rsi]
+        mov     [rdi], rcx
+endif
+        ; At least one write is already done, increment the pointers
+        add     rdi, 8h
+        add     rsi, 8h
+        dec     r8d
+        je      NotInHeapExit
+        ; Now we can do the rest of the writes without checking the heap
+    NextByrefUnchecked:
+        mov     rcx, [rsi]
+        mov     [rdi], rcx
+        add     rdi, 8h
+        add     rsi, 8h
+        dec     r8d
+        jne     NextByrefUnchecked
+    NotInHeapExit:
+        ret
+
+    Exit:
+        ; Increment the pointers before leaving
+        add     rdi, 8h
+        add     rsi, 8h
+        dec     r8d
+        jne     NextByref
```

Comment:
This loop can hang GC suspension.

Reply:
Can you please elaborate? Is it because of shadow gc?

Reply:
We won't be able to suspend the thread for GC while this helper is running. If this helper is caller for large block repeatedly and there is not much managed code running between subsequent invocations of the helper, the GC thread suspension can be very slow or hang.

The same problem existed in `BulkMoveWithWriteBarrier`. It was fixed in https://github.com/dotnet/coreclr/pull/27776 . You should be able to modify the repro from that PR to use InlineArray to hit the hang with this helper.

Reply:
I can limit the max batch size. e.g. 1024 gc slots will be handled as 16 `call CORINFO_HELP_ASSIGN_BYREF_BATCH` batches

Reply:
I would measure the worst-case scenario, e.g. call `MyStruct Test(MyStruct ms) => ms;` in a loop on one thread and measure average GC pause time on a second thread.

Reply:
Thanks, pushed a change to limit the batch size to 256 gcslots (1kb). A test that I used to compare Main vs PR (in Release): https://gist.github.com/EgorBo/270ec3dd078fa69233641907e4773762
(also, checked via `dotnet counters`). E.g. 512 gcslots seemed like too much.

Reply:
I am seeing pause times between 3s and 6s with your repro and your latest change once the tiered compilation kicks in and the code reaches steady state:
```
C:\runtime\artifacts\bin\coreclr\windows.x64.Release>corerun test.exe
6
15
14
14
62
31
47
15
94
63
63
127
62
93
15
3507
3400
3317
6673
15
7669
2388
2817
```

Reply:
@jkotas how does it look for the Main? I tried to be not worth than what we have now. Also, on my machine the longest pause was ~2sec

Reply:
On main, I am seeing 200ms - 500ms for the microbenchmark.

If you create a batch of N, there is going to be N-times fewer good suspension points and so it is going to take N-timer longer for the GC thread suspension to a hit a good suspension point. The math roughly works out for what I am seeing on my machine: 200ms * 16 is about 3s.

Reply:
but why `* 16` ? I my current branch I use 256 as the max batch size. I guess I can lower it to even 16 to fix your numbers?

Reply:
From what I see from the diffs, we're mostly dealing with small values like 2 (most popular) and less than 10

Reply:
Ok, then the math does not work - there is something else going on. (For some reason, I thought that the batch size is 16 in your current change.)

Reply:
If I change the microbenchmark so that the write barrier has to do actual work, I am seeing 15sec - 30sec pause times with your current change: https://gist.github.com/jkotas/d9b48e48935bcc5a6f3dc87db086196d

Reply:
Don't know why, but this benchmark shows better numbers in my branch then in Main